### ISSUE TITLE
Block cursor focus woes

### ISSUE THREAD

**Issue by tohojo** [ch11ng/exwm#437]
I am running exwm on three different machines: My laptop, and two desktops. They all run identical versions of emacs, and share the same .emacs.d - so should be identical setups.

However, on both desktop machines I'm experiencing the same issue that was described in #44 - while on the laptop everything is fine. On the desktop machines I also seem to get flicker when typing in other X apps, so there seems to be some focus issue in general.

How do I debug this?

**Comment by ch11ng** [comment]
This shouldn't be possible with the current implementation I think.  Where is the cursor placed when you type?

> How do I debug this?

You may refer to the wiki to turn on debugging to see if input focus actually shifts during your typing.

**Comment by dakra** [comment]
@tohojo have you seen https://github.com/ch11ng/exwm/issues/435 ?
If the laptop has an older xserver version it would match your description.

**Comment by tohojo** [comment]
Nope, all three run X 1.20 on Arch Linux. I guess I'll try the debug mode and paste the logs...

**Comment by Stebalien** [comment]
I'm also getting this but only on firefox (not other X apps). If I switch from `char` mode to `line` mode, this behaviour goes away.

Emacs Version:
> GNU Emacs 27.0.50 (build 1, x86_64-pc-linux-gnu, X toolkit, Xaw3d scroll bars) of 2018-05-29 (from git)

**Comment by ch11ng** [comment]
@tohojo So there are two issues here, right? One is with conventional Emacs buffers where typing would make Emacs lose focus.  And the other is with X windows (similar to #435 it seems).  For the first one, what I can think of is you may have some global keys (wrongly) set; try removing all global keys from your configuration and see if this issue persists.  As for the second one, aside from Firefox, which apps suffer from this?

**Comment by ch11ng** [comment]
Then it's the same problem with the other issue.

**Comment by tohojo** [comment]
Chris Feng <notifications@github.com> writes:

> Then it's the same problem with the other issue.

Yes, I believe so. The block cursor thing that I actually opened this
issue for is solved by the xmodmap thing. If you have any insights as to
*why* that is, I'd love to hear them; but otherwise, feel free to close this issue :)

**Comment by ch11ng** [comment]
I'm not 100% sure but there may be some global keys wrongly created.  That is, EXWM grabs lots of global keys with actually no binding. It's not clear how that is done.  But the result is, every time you press these keys, the input focus would change to frame's container, so emacs loses focus.

**Comment by ch11ng** [comment]
@tohojo Is this problem still valid for release 0.20? There is a potential fixed included in that release.  Please check it out.

**Comment by tohojo** [comment]
> @tohojo Is this problem still valid for release 0.20? There is a potential fixed included in that release.  Please check it out.

Hmm, depends on what you mean by valid, I suppose? As I commented above, I fixed
the block cursor issue by removing the Xmodmap remapping of keycode 233. If I
re-add that mapping, I still have issues, but I don't need it, so it is not
bothering me.

The other flickering issues are fixed, though :)

**Comment by ch11ng** [comment]
My guess is you have made a global key binding for <kbd>XF86Forward</kbd> somewhere in your Emacs config (`global-set-key` or `exwm-input-global-keys`).  While on your laptop that xmodmap correctly maps the actual forward key to <kbd>XF86Forward</kbd>, on desktop some random key gets mapped (you don't have a forward key on your desktop keyboard, right?).  Please commenting out some of your config and try again, or you may just close this issue if you find it unnecessary to do a further investigation.
